semgrep-core/src/fixing/Autofix_printer.ml

```diff
@@ -0,0 +1,128 @@
+  match lang with
+  | Lang.Python -> Some (new PythonPrinter.printer external_printer)
+  | _ -> None
+
+let original_source_of_ast source any =
+  let* start, end_ = Visitor_AST.range_of_any_opt any in
+  let starti = start.Parse_info.charpos in
+  let _, _, endi = Parse_info.get_token_end_info end_ in
+  let len = endi - starti in
+  let str = String.sub source starti len in
+  Some str
+
+(* Add each metavariable value to the lookup table so that it can be identified
+ * during printing *)
+let add_metavars tbl metavars =
+  List.iter
+    (fun (_, mval) ->
+      let any = MV.mvalue_to_any mval in
+      Hashtbl.replace tbl any `Metavar;
+      (* For each metavariable binding that is a list of things, we need to
+       * iterate through and add each item in the list to the table as well.
+       *
+       * For example, if `$...BAR` is bound to the separate arguments `1` and
+       * `2`, in the example below, then the complete argument list `(1, 2)`
+       * would never appear in the resulting AST that we attempt to print.
+       * Despite that, we would like to reuse the original text for `1` and `2`.
+       *
+       * foo($...BAR, 5) -> foo(1, 2, 5)
+       *
+       * We don't need to recurse any deeper, because individual list items are
+       * the smallest components of a metavariable value that will be lifted
+       * into the resulting AST.
+       * *)
+      match mval with
+      | MV.Args args ->
+          List.iter (fun arg -> Hashtbl.replace tbl (Ar arg) `Metavar) args
+      (* TODO iterate through other metavariable values that are lists *)
+      | _ -> ())
+    metavars
+
+(* Add each AST node to the lookup table so that it can be identified during
+ * printing *)
+let add_ast_nodes tbl ast =
+  let visitor =
+    Visitor_AST.(
+      mk_visitor
+        {
+          default_visitor with
+          kargument =
+            (fun (k, _) arg ->
+              Hashtbl.replace tbl (Ar arg) `Pattern;
+              k arg)
+            (* TODO visit every node that is part of AST_generic.any *);
+        })
+  in
+  visitor ast
+
+let make_external_printer ~metavars ~target_contents ~fix_pattern_ast
+    ~fix_pattern : AST_generic.any -> Immutable_buffer.t option =
+  (* TODO Avoid polymorphic hash function. It takes into account parts of the
+   * AST that we probably don't want to use for equality. Most likely, we should
+   * use the hash function generated by ppx deriving, which already ignores
+   * things as appropriate. For now, this seems to work fine, and anyway,
+   * failing to look up a node is not catastrophic: it will only mean that in
+   * some cases we may fail to print. *)
+  let tbl = Hashtbl.create 8 in
+  add_metavars tbl metavars;
+  add_ast_nodes tbl fix_pattern_ast;
+  fun any ->
+    let* node = Hashtbl.find_opt tbl any in
+    let* str =
+      match node with
+      | `Metavar -> original_source_of_ast (Lazy.force target_contents) any
```

Comment:
It's really counterintuitive to me that "metavar" is the one we can use the original source of the ast for but "pattern" is for the fix pattern? I'm still missing something. I'll leave this file now with my scattered comments. tldr; explain more in the file header with examples

Reply:
This is because the metavar bindings contain the AST nodes that those metavars matched in the original target file. So, when printing those nodes, we need to get them from the target file.

In contrast, when printing nodes from the fix pattern, we need to lift text from the fix pattern in the rule.

I'll add some comments and think about how I can improve the naming here to make it clearer.